src/shared/lib/URLWrapper.ts

```diff
@@ -0,0 +1,51 @@
+import {autorun, computed, extendObservable, intercept, IObservableObject, observable} from "mobx";
+import ExtendedRouterStore from "./ExtendedRouterStore";
+
+export type Property<T> = {
+    name: keyof T,
+    persisted: boolean
+};
+
+export default class URLWrapper<QueryParamsType> {
+    public query:Partial<QueryParamsType>;
+
+    constructor(
+        protected routing:ExtendedRouterStore,
+        protected properties:Property<QueryParamsType>[]
+    ) {
+        const initValues:Partial<QueryParamsType> = {};
+        for (const property of properties) {
+            initValues[property.name] = undefined;
+        }
+        this.query = observable<Partial<QueryParamsType>>(initValues);
+
+        intercept(this.query, change=>{
+            if (change.newValue === this.query[change.name as keyof QueryParamsType]) {
+                // if same value, cancel change to prevent unnecessary changing data
+                return null;
+            } else {
+                return change;
+            }
+        });
+        autorun(()=>{
+            const query = routing.query as QueryParamsType;
+            for (const property of properties) {
+                this.query[property.name] = query[property.name];
+            }
+        });
+    }
+
+    public updateQuery(query:Partial<QueryParamsType>) {
```

Comment:
what's thinking of putting this here?  i mean as opposed to just letting code use router directly.    updateRoute has optional params to clear state btw.  if we decide to keep this maybe we need that here too?

Reply:
my thinking was to put getting and setting in the same place - I thought it might be a bit confusing for a user to get using wrapper[param] and then set using routing.updateRoute. Wouldnt be opposed to adding those optional params here.